subprojects/dependency-management/src/main/java/org/gradle/internal/component/external/model/ImmutableCapability.java

```diff
@@ -16,19 +16,21 @@
 package org.gradle.internal.component.external.model;
 
 import com.google.common.base.Objects;
-import org.gradle.api.capabilities.CapabilityDescriptor;
 
-public class ImmutableCapability implements CapabilityDescriptor {
+public class ImmutableCapability implements CapabilityInternal {
+
     private final String group;
     private final String name;
     private final String version;
     private final int hashCode;
+    private final String cachedId;
 
     public ImmutableCapability(String group, String name, String version) {
         this.group = group;
         this.name = name;
         this.version = version;
         this.hashCode = Objects.hashCode(group, name, version);
+        this.cachedId = group + ":" + name;
```

Comment:
Note to the reviewer: it might look strange, but this turned out to be the fastest of several experiments, including:

- using `ModuleIdentifier` (initial implementation)
- using `ModuleIdentifier` through `ImmutableModuleIdentifierFactory` (for interning)
- using a 2-level map (by group, then by name)
- using an interned string for the `cachedId` (interning turned out to cost as much as what we gain from faster checks in maps)

Reply:
It won't hurt to add a Javadoc to `ImmutableCapability` that includes these explanations about the implementation decisions.

You can also add a note about computing the hashCode up front (I guess it is always needed, so there is no point in computing it lazily).